napari/_app_model/_app.py

```diff
@@ -25,22 +54,159 @@ def __init__(self, app_name=APP_NAME) -> None:
+        #     action_manager wants "napari:toggle_theme"
+        # so this hack works if we keep the prefix and suffix the same when
+        # porting to app-model
+        prefix, *_group, command = action.id.split(":")
+
+        if isinstance(action.callback, GeneratorCallback):
+
+            def _callback(*args, **kwargs):
+                self.get_app().commands.execute_command(action.id).result()
+                yield
+                self.get_app().commands.execute_command(action.id).result()
+
+        else:
+
+            def _callback(*args, **kwargs):
+                return (
+                    self.get_app().commands.execute_command(action.id).result()
+                )
+
+        _callback.__name__ = command
+
+        action_manager.register_action(
+            name=f"{prefix}:{command}",
+            command=_callback,
+            description=action.title,
+            keymapprovider=keymapprovider,
+            repeatable=isinstance(action, RepeatableAction),
+        )
+
     @classmethod
     def get_app(cls, app_name: str = APP_NAME) -> NapariApplication:
         return Application.get_app(app_name) or cls()  # type: ignore[return-value]
 
+    def register_action(self, action: Action) -> Callable[[], None]:
+        dispose = super().register_action(action)
+        if isinstance(action, RepeatableAction) and action.repeatable:
+            self.set_action_repeatable(action.id, True)
+
+            def _dispose():
+                dispose()
+                self._repeatable_actions.remove(action.id)
+
+            return _dispose
+        return dispose
+
+    def set_action_repeatable(self, action_id: str, repeatable: bool):
+        """Registers an action as repeatable or not.
+
+        Parameters
+        ----------
+        action_id : str
+            Unique identifier of the action.
+        repeatable : bool
+            Whether or not to set the action as repeatable.
+        """
+        if action_id not in self.commands:
+            raise ValueError(f'Command {id!r} not registered')
+
+        if repeatable:
+            self._repeatable_actions.add(action_id)
+        else:
+            self._repeatable_actions.discard(action_id)
+
+    def action_is_repeatable(self, action_id: str) -> bool:
+        """Determines if an action is repeatable.
+
+        Parameters
+        ----------
+        action_id : str
+            Unique identifier of the action.
+
+        Returns
+        -------
+        bool
+            Whether the action is repeatable or not.
+        """
+        return action_id in self._repeatable_actions
+
+    def _on_shortcuts_changed(self, _):
+        self.keybindings.discard_entries(KeyBindingWeights.USER)
+        for entry in get_settings().shortcuts.shortcuts:
+            self.keybindings.register_keybinding_rule(
+                entry.command,
+                KeyBindingRule(
+                    primary=KeyBinding.from_str(entry.key),
```

Comment:
Seems like adding a call to the `to_int` method here can also help with `pydantic` validation errors. So something like:

```suggestion
                    primary=KeyBinding.from_str(entry.key).to_int(),
```